src/rest/resource.cpp

```diff
@@ -800,6 +809,256 @@ void Resource::DatasetPending(const Request &aRequest, Response &aResponse) cons
+    std::vector<otJoinerInfo> joinerTable;
+    std::string               joinerJson;
+    std::string               errorCode;
+
+    while (otCommissionerGetNextJoinerInfo(mInstance, &iter, &joinerInfo) == OT_ERROR_NONE)
+    {
+        joinerTable.push_back(joinerInfo);
+    }
+
+    joinerJson = Json::JoinerTable2JsonString(joinerTable);
+    aResponse.SetBody(joinerJson);
+    errorCode = GetHttpStatus(HttpStatusCode::kStatusOk);
+    aResponse.SetResponsCode(errorCode);
+}
+
+void Resource::AddJoiner(const Request &aRequest, Response &aResponse) const
+{
+    otbrError           error   = OTBR_ERROR_NONE;
+    otError             errorOt = OT_ERROR_NONE;
+    std::string         errorCode;
+    otJoinerInfo        joiner;
+    const otExtAddress *addrPtr                         = nullptr;
+    const uint8_t       emptyArray[OT_EXT_ADDRESS_SIZE] = {0};
+
+    VerifyOrExit(otCommissionerGetState(mInstance) == OT_COMMISSIONER_STATE_ACTIVE, error = OTBR_ERROR_INVALID_STATE);
+
+    VerifyOrExit(Json::JsonJoinerInfoString2JoinerInfo(aRequest.GetBody(), joiner), error = OTBR_ERROR_INVALID_ARGS);
+
+    addrPtr = &joiner.mSharedId.mEui64;
+    if (memcmp(&joiner.mSharedId.mEui64, emptyArray, OT_EXT_ADDRESS_SIZE) == 0)
+    {
+        addrPtr = nullptr;
+    }
+
+    if (joiner.mType == OT_JOINER_INFO_TYPE_DISCERNER)
+    {
+        errorOt = otCommissionerAddJoinerWithDiscerner(mInstance, &joiner.mSharedId.mDiscerner, joiner.mPskd.m8,
+                                                       joiner.mExpirationTime);
+    }
+    else
+    {
+        errorOt = otCommissionerAddJoiner(mInstance, addrPtr, joiner.mPskd.m8, joiner.mExpirationTime);
+    }
+    VerifyOrExit(errorOt == OT_ERROR_NONE, error = OTBR_ERROR_OPENTHREAD);
+
+exit:
+    if (error == OTBR_ERROR_NONE)
+    {
+        errorCode = GetHttpStatus(HttpStatusCode::kStatusOk);
+        aResponse.SetResponsCode(errorCode);
+    }
+    else if (error == OTBR_ERROR_INVALID_STATE)
+    {
+        ErrorHandler(aResponse, HttpStatusCode::kStatusConflict);
+    }
+    else if (error == OTBR_ERROR_INVALID_ARGS)
+    {
+        ErrorHandler(aResponse, HttpStatusCode::kStatusBadRequest);
+    }
+    else if (error == OTBR_ERROR_OPENTHREAD)
+    {
+        if (errorOt == OT_ERROR_INVALID_ARGS)
+        {
+            ErrorHandler(aResponse, HttpStatusCode::kStatusBadRequest);
+        }
+        else if (errorOt == OT_ERROR_NO_BUFS)
+        {
+            ErrorHandler(aResponse, HttpStatusCode::kStatusInsufficientStorage);
+        }
+        else
+        {
+            ErrorHandler(aResponse, HttpStatusCode::kStatusInternalServerError);
+        }
+    }
+    else
+    {
+        ErrorHandler(aResponse, HttpStatusCode::kStatusInternalServerError);
+    }
+}
+
+void Resource::RemoveJoiner(const Request &aRequest, Response &aResponse) const
+{
+    otbrError         error = OTBR_ERROR_NONE;
+    std::string       errorCode;
+    otExtAddress      eui64;
+    otExtAddress     *addrPtr   = nullptr;
```

Comment:
I think this `addrPtr` is not necessary.

Reply:
it's there to provide a NULL argument to remove all if the wildcard `*` was given or to point to the eui64 if the wildcard was not given

Reply:
Oh, I missed that. The current code looks good, thanks!